File: baekjoon/src/tags/dfsAndBfs/No1260.java
package tags.dfsAndBfs;

import java.io.*;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;
import java.util.StringTokenizer;

/**
 * https://www.acmicpc.net/problem/1260
 */
public class No1260 {
    public static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    public static int nodes;
    public static int connections;
    public static int startAt;
    public static boolean[] isVisited;
    public static int[][] map;

    public static void setMap() throws IOException {
        StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
        nodes = Integer.parseInt(tokenizer.nextToken());
        connections = Integer.parseInt(tokenizer.nextToken());
        startAt = Integer.parseInt(tokenizer.nextToken());

        map = new int[nodes + 1][nodes + 1];
        for (int i = 0; i < connections; i++) {
            tokenizer = new StringTokenizer(reader.readLine());
            int source = Integer.parseInt(tokenizer.nextToken());
            int destination = Integer.parseInt(tokenizer.nextToken());
            map[source][destination] = map[destination][source] = 1;
        }
    }

    public static Stack<Integer> stack = new Stack<>();

    public static Queue<Integer> queue = new LinkedList<>();

    public static void dfs(int startAt) throws IOException {
        stack.push(startAt);
        isVisited[startAt] = true;
        writer.write(startAt + " ");

        for (int i = 1; i <= nodes; i++) {
            if (map[startAt][i] == 1 && !isVisited[i])
                dfs(i);
        }
    }

    public static void bfs(int startAt) throws IOException {
        queue.add(startAt);
        isVisited[startAt] = true;

        while (!queue.isEmpty()) {
            int number = queue.poll();
            writer.write(number + " ");

            for (int i = 1; i <= nodes; i++) {
                if (map[number][i] == 1 && !isVisited[i]) {
                    queue.add(i);
                    isVisited[i] = true;
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        setMap();

        isVisited = new boolean[nodes + 1];
        dfs(startAt);
        writer.newLine();

        isVisited = new boolean[nodes + 1];
        bfs(startAt);
        writer.flush();
    }
}

File: baekjoon-classification/src/dfs/No1260.kt
package dfs

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

/**
 * https://www.acmicpc.net/problem/1260
 */
private val reader = BufferedReader(InputStreamReader(System.`in`))
private val writer = BufferedWriter(OutputStreamWriter(System.out))

fun main() {
    var tokenizer = StringTokenizer(reader.readLine())
    val edges = tokenizer.nextToken().toInt()
    val nodes = tokenizer.nextToken().toInt()
    val startIndex = tokenizer.nextToken().toInt()

    val graph = Array(edges + 1) { IntArray(edges + 1) }
    repeat(nodes) {
        tokenizer = StringTokenizer(reader.readLine())
        val row = tokenizer.nextToken().toInt()
        val col = tokenizer.nextToken().toInt()
        graph[row][col] = 1
        graph[col][row] = 1
    }

    var isVisited = BooleanArray(edges + 1)
    dfs(graph, isVisited, startIndex)
    writer.newLine()

    val queue = LinkedList<Int>()
    isVisited = BooleanArray(edges + 1)
    bfs(queue, graph, isVisited, startIndex)
    writer.flush()
}

private fun dfs(graph: Array<IntArray>, isVisited: BooleanArray, index: Int) {
    isVisited[index] = true
    writer.write("$index ")

    for (i in 1 until isVisited.size) {
        if (!isVisited[i] && graph[index][i] == 1)
            dfs(graph, isVisited, i)
    }
}

private fun bfs(queue: Queue<Int>, graph: Array<IntArray>, isVisited: BooleanArray, index: Int) {
    isVisited[index] = true
    queue.add(index)

    while (queue.isNotEmpty()) {
        val newIndex = queue.poll()
        writer.write("$newIndex ")

        for (i in 1 until isVisited.size) {
            if (!isVisited[i] && graph[newIndex][i] == 1) {
                isVisited[i] = true
                queue.add(i)
            }
        }
    }
}